File: ViewModelFramework/Android/basics/AbstractViewModel.java
package de.myparcelservice.app.ViewModels.basics;


import java.util.ArrayList;
import java.util.List;

import de.myparcelservice.app.OSAPIInterfaces.IAsyncDelegate;
import de.myparcelservice.app.OSAPIInterfaces.IAsyncTask;
import de.myparcelservice.app.OSAPIInterfaces.IAsyncTaskCreator;

/**
 * @sharpen.ignore
 * Diese Klasse wird bei der Sprachkonversion nach C# ausgelassen.
 * Grund dafür ist die in C# mit Events realisierte Anbindung an die UI
 * @param <TModel>
 */
public abstract class AbstractViewModel<TModel>
{

	private int _initializationProgress;
	private List<IViewModelObserver> _observers;
	protected TModel _model;
    private IAsyncTask<TModel,Integer,Void> _asyncTask;
    private IAsyncTaskCreator _asyncTaskCreator;

    @SuppressWarnings("unchecked")
	public AbstractViewModel(TModel model, IAsyncTaskCreator asyncTaskCreator)
	{
        _asyncTaskCreator = asyncTaskCreator;
        _initializationProgress=0;
		_model=model;
		_observers = new ArrayList<IViewModelObserver>();

		
	}
protected final void startInitialization()
{
    if(_asyncTaskCreator!=null)
    {
        _asyncTask = _asyncTaskCreator.<TModel, Integer, Void>createAsyncTask(getInitializer());
        _asyncTask.Execute(_model);
    }
}


    IAsyncDelegate<TModel,Integer,Void> getInitializer()
    {
        return new IAsyncDelegate<TModel, Integer, Void>()
        {
            @Override
            public Void doInBackground(TModel... tModels)
            {
                initializeAsynchronously();
                return null;
            }


            @Override
            public void onProgressUpdate(Integer... values)
            {
                _initializationProgress = values[0];
                RaisePropertyChanged("InitializationProgress");
            }

            @Override
            public  void onPostExecute(Void einVoid)
            {
                onInitializationDone();
            }

            @Override
            public void onPreExecute()
            {

            }

        };
    }

	public int GetInitializationProgress()
	{
		return _initializationProgress;
	}
	@Override
	public TModel GetModel()
	{
		return _model;
	}

	@Override
	public void AddObserver(IViewModelObserver newObserver)
	{
		_observers.add(newObserver);
	}

	public void removeObserver(IViewModelObserver observerToRemove)
	{
		if (_observers.contains(observerToRemove))
		{
			_observers.remove(observerToRemove);
		}
	}
	
	protected void RaisePropertyChanged(String propertyName)
	{
		for(IViewModelObserver observer: _observers)
		{
			observer.Update(this, propertyName);
		}
	}

	@Override
	public void SetModel(TModel model)
	{
		_model = model;
	}

	

	protected abstract void initializeAsynchronously();
	
	protected void updateInitializationProgress(int progress)
	{
		_asyncTask.PublishProgress(progress);
	}

	protected abstract void onInitializationDone();

}

File: ViewModelFramework/Windows Phone/basics/AbstractViewModel.cs
using System;
using System.Collections.Generic;
using DE.Myparcelservice.App.OSAPIInterfaces;
using System.ComponentModel;

namespace DE.Myparcelservice.App.ViewModels.Basics
{
    public abstract class AbstractViewModel<TModel> : INotifyPropertyChanged
    {
        private int _initializationProgress;

        public event PropertyChangedEventHandler PropertyChanged;


        protected internal TModel _model;

        private IAsyncTask<TModel, int, object> _asyncTask;
        private IAsyncTaskCreator _asyncTaskCreator;

        public AbstractViewModel(TModel model, IAsyncTaskCreator asyncTaskCreator)
        {
            _asyncTaskCreator = asyncTaskCreator;
            _initializationProgress = 0;
            _model = model;

            
        }
        protected void startInitialization()
{
        if (_asyncTaskCreator != null)
            {
                _asyncTask = _asyncTaskCreator.CreateAsyncTask<TModel, int, object>(GetInitializer
                    ());
                _asyncTask.Execute(_model);
            }
}

        internal virtual IAsyncDelegate<TModel, int, object> GetInitializer()
        {
            return new _IAsyncDelegate_34(this);
        }

        private sealed class _IAsyncDelegate_34 : IAsyncDelegate<TModel, int, object>
        {
            public _IAsyncDelegate_34(AbstractViewModel<TModel> _enclosing)
            {
                this._enclosing = _enclosing;
            }

            public object DoInBackground(params TModel[] tModels)
            {
                this._enclosing.InitializeAsynchronously();
                return null;
            }

            public void OnProgressUpdate(params int[] values)
            {
                this._enclosing._initializationProgress = values[0];
                this._enclosing.RaisePropertyChanged("InitializationProgress");
            }

            public void OnPostExecute(object einVoid)
            {
                this._enclosing.OnInitializationDone();
            }

            public void OnPreExecute()
            {
            }

            private readonly AbstractViewModel<TModel> _enclosing;
        }

        public virtual int GetInitializationProgress()
        {
            return _initializationProgress;
        }

        public virtual TModel GetModel()
        {
            return _model;
        }



        protected internal virtual void RaisePropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
            {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }

        public virtual void SetModel(TModel model)
        {
            _model = model;
        }

        protected internal abstract void InitializeAsynchronously();

        protected internal virtual void UpdateInitializationProgress(int progress)
        {
            _asyncTask.PublishProgress(progress);
        }

        protected internal abstract void OnInitializationDone();

    }
}
